### PR TITLE
Feature/6529 check in selection in submission flow

### PR DESCRIPTION
**Pull request by marcussc**
## Description
During Submission the User has to select the TraceLocations/Checkins that shall be submitted.

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-6529

## Screenshots
Screen Shot 2021-04-18 at 10 55 13
Screen Shot 2021-04-18 at 10 55 07
Screen Shot 2021-04-18 at 10 55 00
Screen Shot 2021-04-18 at 10 54 56
Screen Shot 2021-04-18 at 10 54 48
Screen Shot 2021-04-18 at 10 54 42
Screen Shot 2021-04-18 at 10 54 24
Screen Shot 2021-04-18 at 10 53 35
Screen Shot 2021-04-18 at 10 53 26
Screen Shot 2021-04-18 at 10 53 08

### CODE REVIEW

**Review by Ein-Tim — COMMENTED** [review]
@marcussc Great work! And this on the weekend! 🚀 Thank you! ❤️

One question:
Do we really want to use "in Ihrer Nähe waren" & "who checked in near you"?
All users who checked in will be warned, regardless if they were near me or not.
Pinging @janetback here 🙂

**Review by nickguendling — APPROVED** [review]

File: src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings [hunk 1]
```diff
@@ -2252,3 +2252,23 @@ Die Angaben werden statistisch ausgewertet, sie werden nicht zu einem Profil ges
...
+
+"Submission_Checkins_Description" = "Teilen Sie Ihre Check-ins, um andere zu warnen, die in Ihrer Nähe waren. Ihre Identität bleibt geheim.";
+
+"Submission_Checkins_SelectAll" = "Alle auswählen";
+
+"Submission_Checkins_Continue" = "Weiter";
+
+"Submission_Checkins_Skip" = "Überspringen";
+
+"Submission_Checkins_Alert_Title" = "Sind Sie sicher, dass Sie Ihre Check-Ins nicht teilen wollen?";
```
Comment (janetback) [inline src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings:2268]: Hi, could you please change to small 'i' -> "Check-ins". Thanks.

File: src/xcode/ENA/ENA/Resources/Localization/en.lproj/Localizable.strings [hunk 2]
```diff
@@ -2193,3 +2193,21 @@
...
+
+"Submission_Checkins_Description" = "Warn others who checked in near you. Your personal data will not be shared.";
+
+"Submission_Checkins_SelectAll" = "Select all";
+
+"Submission_Checkins_Continue" = "Submit";
+
+"Submission_Checkins_Skip" = "Skip";
+
+"Submission_Checkins_Alert_Title" = "Are you sure you don’t want to share your Check-ins?";
```
Comment (janetback) [inline src/xcode/ENA/ENA/Resources/Localization/en.lproj/Localizable.strings:2207]: Could you please change to small 'c' -> check-ins. Thanks.

File: src/xcode/ENA/ENA/Resources/Localization/en.lproj/Localizable.strings [hunk 3]
```diff
@@ -2193,3 +2193,21 @@
...
+
+"Submission_Checkins_Alert_Message" = "Skipping will not warn any contacts who checked-in near you.";
```
Comment (janetback) [inline src/xcode/ENA/ENA/Resources/Localization/en.lproj/Localizable.strings:2209]: please take out the hyphen -> checked in near you.

**Review by janetback — CHANGES_REQUESTED** [review]
Hi, there are a few small changes need to the text (see comments). Thanks.

**Comment by janetback** [comment]
> @marcussc Great work! And this on the weekend! 🚀 Thank you! ❤️
> 
> One question:
> Do we really want to use "in Ihrer Nähe waren" & "who checked in near you"?
> All users who checked in will be warned, regardless if they were near me or not.
> Pinging @janetback here 🙂

Hi @Ein-Tim - I was instructed to stick to the English text here ('near you'), but have already raised the issue and asked for this to be checked.

**Comment by Ein-Tim** [comment]
Thanks @janetback 

Even if it doesn't get changed, here my suggestion: 
"Warn others who checked in near you. Your personal data will not be shared." 
to 
"Warn others who checked in with you. Your personal data will not be shared." 
& 
"Skipping will not warn any contacts who checked-in near you."
to 
"Skipping will not warn any contacts who checked in with you."

And German: 
"Teilen Sie Ihre Check-ins, um andere zu warnen, die in Ihrer Nähe waren. Ihre Identität bleibt geheim." 
to
"Teilen Sie Ihre Check-ins, um andere zu warnen, die mit Ihnen eingecheckt waren. Ihre Identität bleibt geheim."
&
"Dadurch werden andere nicht gewarnt, die in Ihrer Nähe waren."
to
"Dadurch werden andere nicht gewarnt, die mit Ihnen eingecheckt waren."

File: src/xcode/ENA/ENA/Resources/Localization/en.lproj/Localizable.strings (hunk 3 above)
Comment (marcussc) [reply src/xcode/ENA/ENA/Resources/Localization/en.lproj/Localizable.strings:2209]: Done

File: src/xcode/ENA/ENA/Resources/Localization/en.lproj/Localizable.strings (hunk 2 above)
Comment (marcussc) [reply src/xcode/ENA/ENA/Resources/Localization/en.lproj/Localizable.strings:2207]: Done

File: src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings (hunk 1 above)
Comment (marcussc) [reply src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings:2268]: Done

**Review by janetback — APPROVED** [review]
UA approved

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionCheckins/Cells/ExposureSubmissionCheckinCellModel.swift [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	
+	var selected: Bool = false {
+		didSet {
+			cellIsSelected = selected
+			checkmarkImage = selected ? UIImage(named: "Checkin_Checkmark_Selected") : UIImage(named: "Checkin_Checkmark_Unselected")
+		}
+	}
+
+	@OpenCombine.Published var cellIsSelected: Bool = false
+	@OpenCombine.Published var checkmarkImage = UIImage(named: "Checkin_Checkmark_Unselected")
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionCheckins/Cells/ExposureSubmissionCheckinCellModel.swift:44]:
```suggestion
	@OpenCombine.Published private(set)  var cellIsSelected: Bool = false
	@OpenCombine.Published private(set) var checkmarkImage = UIImage(named: "Checkin_Checkmark_Unselected")
```

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionCheckins/Cells/ExposureSubmissionCheckinCellModel.swift [hunk 5]
```diff
@@ -0,0 +1,46 @@
+////
+// 🦠 Corona-Warn-App
+//
+
+import UIKit
+import OpenCombine
+
+class ExposureSubmissionCheckinCellModel {
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionCheckins/Cells/ExposureSubmissionCheckinCellModel.swift:8]:
```suggestion
class ExposureSubmissionCheckInCellModel {
```

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionCheckins/Cells/ExposureSubmissionCheckinDescriptionCellModel.swift [hunk 6]
```diff
@@ -0,0 +1,13 @@
+////
+// 🦠 Corona-Warn-App
+//
+
+import Foundation
+
+struct ExposureSubmissionCheckinDescriptionCellModel {
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionCheckins/Cells/ExposureSubmissionCheckinDescriptionCellModel.swift:7]:
```suggestion
struct ExposureSubmissionCheckInDescriptionCellModel {
```

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionCheckins/Cells/ExposureSubmissionCheckinDescriptionTableViewCell.swift [hunk 7]
```diff
@@ -0,0 +1,47 @@
+////
+// 🦠 Corona-Warn-App
+//
+
+import UIKit
+
+class ExposureSubmissionCheckinDescriptionTableViewCell: UITableViewCell, ReuseIdentifierProviding {
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionCheckins/Cells/ExposureSubmissionCheckinDescriptionTableViewCell.swift:7]:
```suggestion
class ExposureSubmissionCheckInDescriptionTableViewCell: UITableViewCell, ReuseIdentifierProviding {
```

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionCheckins/Cells/ExposureSubmissionCheckinTableViewCell.swift [hunk 8]
```diff
@@ -0,0 +1,130 @@
+////
+// 🦠 Corona-Warn-App
+//
+
+import UIKit
+import OpenCombine
+
+class ExposureSubmissionCheckinTableViewCell: UITableViewCell, ReuseIdentifierProviding {
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionCheckins/Cells/ExposureSubmissionCheckinTableViewCell.swift:8]:
```suggestion
class ExposureSubmissionCheckInTableViewCell: UITableViewCell, ReuseIdentifierProviding {
```

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionCheckins/ExposureSubmissionCheckinsViewController.swift [hunk 9]
```diff
@@ -0,0 +1,139 @@
+////
+// 🦠 Corona-Warn-App
+//
+
+import UIKit
+import OpenCombine
+
+class ExposureSubmissionCheckinsViewController: UITableViewController, DismissHandling, FooterViewHandling {
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionCheckins/ExposureSubmissionCheckinsViewController.swift:8]:
```suggestion
class ExposureSubmissionCheckInsViewController: UITableViewController, DismissHandling, FooterViewHandling {
```

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionCheckins/ExposureSubmissionCheckinsViewController.swift [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+
+	// MARK: - Init
+	
+	init(checkins: [Checkin], onCompletion: @escaping ([Checkin]) -> Void, onSkip: @escaping () -> Void, onDismiss: @escaping () -> Void) {
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionCheckins/ExposureSubmissionCheckinsViewController.swift:12]:
```suggestion
	init(
	checkIns: [Checkin],
	onCompletion: @escaping ([Checkin]) -> Void,
	onSkip: @escaping () -> Void,
	onDismiss: @escaping () -> Void) 
	{
```

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionCheckins/ExposureSubmissionCheckinsViewController.swift [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+	
+	override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+		switch ExposureSubmissionCheckinsViewModel.Section(rawValue: indexPath.section) {
+		case .description:
+			return
+		case .checkins:
+			viewModel.toogleSelection(at: indexPath.row)
+			return
+		default:
+			fatalError("Invalid section")
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionCheckins/ExposureSubmissionCheckinsViewController.swift:112]: better log.error instead fo a fatal error here?

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionCheckins/ExposureSubmissionCheckinsViewModel.swift [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+	// MARK: - Internal
+	
+	enum Section: Int, CaseIterable {
+		case description
+		case checkins
+	}
+	
+	let title = AppStrings.ExposureSubmissionCheckins.title
+	let checkinCellModels: [ExposureSubmissionCheckinCellModel]
+	@OpenCombine.Published var continueEnabled: Bool = false
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionCheckins/ExposureSubmissionCheckinsViewModel.swift:27]:
```suggestion
	@OpenCombine.Published private(set) var continueEnabled: Bool = false
```

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionCheckins/ExposureSubmissionCheckinsViewModel.swift [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+	}
+	
+	func numberOfRows(in section: Int) -> Int {
+		switch Section(rawValue: section) {
+		case .description:
+			return 1
+		case .checkins:
+			return checkinCellModels.count
+		case .none:
+			fatalError("Invalid section")
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionCheckins/ExposureSubmissionCheckinsViewModel.swift:46]: Log.error and return 0 - might be better than a fatal error?

**Review by KaiTeuber — APPROVED** [review]
We should have a consensus in the team how we want to write `checkin` ? `checkIn` or`checkin` - currently we have a mixed bag in the code.

**Comment by marcussc** [comment]
> We should have a consensus in the team how we want to write `checkin` ? `checkIn` or`checkin` - currently we have a mixed bag in the code.

@KaiTeuber Yeah, I resolved your suggested changes regarding this Issue. @nickguendling and I aligned the naming with the Type `Checkin`

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionCheckins/ExposureSubmissionCheckinsViewController.swift (hunk 11 above)
Comment (marcussc) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionCheckins/ExposureSubmissionCheckinsViewController.swift:112]: Done

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionCheckins/ExposureSubmissionCheckinsViewModel.swift (hunk 13 above)
Comment (marcussc) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionCheckins/ExposureSubmissionCheckinsViewModel.swift:46]: Done

**Pull request merged**